plugins/glustershd/transactions.go

```diff
@@ -1,71 +1,132 @@
+func runGlfshealBin(c transaction.TxnCtx, volname string, args []string) error {
+	var out bytes.Buffer
+	var buffer bytes.Buffer
 
-const (
-	actionStart actionType = iota
-	actionStop
-)
+	buffer.WriteString(fmt.Sprintf("%s", volname))
+	for _, arg := range args {
+		buffer.WriteString(fmt.Sprintf(" %s", arg))
+	}
 
-func selfhealdAction(c transaction.TxnCtx, action actionType) error {
-	glustershDaemon, err := newGlustershd()
+	args = strings.Fields(buffer.String())
+	path, err := exec.LookPath("glfsheal")
 	if err != nil {
+		c.Logger().WithError(err).Error("Couldn't find glfsheal binary.")
 		return err
 	}
-	switch action {
-	case actionStart:
-		err = daemon.Start(glustershDaemon, true)
-		if err == errors.ErrProcessAlreadyRunning {
-			c.Logger().WithError(err).Error("Seld Heal Daemon is already running.")
-			return nil
-		}
-	case actionStop:
-		isVolRunning, err := volume.AreReplicateVolumesRunning()
-		if err != nil {
-			c.Logger().WithError(err).Error("Failed to get volinfo. Etcd server might be down.")
-			return err
-		}
-		if !isVolRunning {
-			err = daemon.Stop(glustershDaemon, true)
-			if err != nil {
-				c.Logger().WithError(err).Error("Failed to stop self heal daemon.")
-				return err
-			}
-		}
+
+	cmd := exec.Command(path, args...)
+	cmd.Stdout = &out
+	err = cmd.Run()
+	if err != nil {
+		c.Logger().WithError(err).Error("Failed to execute Glfsheal Binary")
+		return err
+	}
+	if err := c.Set("stdout", out.String()); err != nil {
+		c.Logger().WithError(err).WithField(
+			"key", "stdout").Debug("Failed to set key in transaction context.")
+		return err
 	}
-	return err
+	return nil
 }
 
-func txnSelfHealStart(c transaction.TxnCtx) error {
-	return selfhealdAction(c, actionStart)
-}
+func txnHealInfo(c transaction.TxnCtx) error {
+	var volname string
+	var args []string
 
-func txnSelfHealStop(c transaction.TxnCtx) error {
-	return selfhealdAction(c, actionStop)
-}
+	if err := c.Get("volname", &volname); err != nil {
+		c.Logger().WithError(err).WithField(
+			"key", "volname").Debug("Failed to get key from transaction context.")
+		return err
+	}
 
-func txnSelfHealdUndo(c transaction.TxnCtx) error {
-	var oldvolinfo volume.Volinfo
-	if err := c.Get("oldvolinfo", &oldvolinfo); err != nil {
+	if err := c.Get("option", &args); err != nil {
+		c.Logger().WithError(err).WithField(
+			"key", "option").Debug("Failed to get key from transaction context.")
 		return err
 	}
-	if err := volume.AddOrUpdateVolumeFunc(&oldvolinfo); err != nil {
+
+	return runGlfshealBin(c, volname, args)
+}
+
+func txnValidateVolNameType(c transaction.TxnCtx) error {
+
+	var volname string
+
+	if err := c.Get("volname", &volname); err != nil {
 		c.Logger().WithError(err).WithField(
-			"volume", oldvolinfo.Name).Debug("storeVolume: failed to store volume info")
+			"key", "volname").Debug("Failed to get key from transaction context.")
 		return err
 	}
-	if err := volgen.Generate(); err != nil {
+	// Validate volume existence
```

Comment:
These need be step functions. You can simply do these validations in the handler right after taking locks.